infrastructure/storage/src/key_val_store/hmap_database.rs

```diff
@@ -0,0 +1,245 @@
+        bincode::serialize_into(&mut value_buf, value)
+            .map_err(|e| KeyValStoreError::SerializationError(e.to_string()))?;
+
+        self.db
+            .write()
+            .map_err(|_| KeyValStoreError::PoisonedAccess)?
+            .insert(key.as_lmdb_bytes().to_vec(), value_buf);
+        Ok(())
+    }
+
+    /// Get a value from the key-value database. The retrieved value is deserialized from bincode into `V`
+    pub fn get<K, V>(&self, key: &K) -> Result<Option<V>, KeyValStoreError>
+    where
+        K: AsLmdbBytes + ?Sized,
+        for<'t> V: serde::de::DeserializeOwned, // read this as, for *any* lifetime, t, we can convert a [u8] to V
+    {
+        match self
+            .db
+            .read()
+            .map_err(|_| KeyValStoreError::PoisonedAccess)?
+            .get(&key.as_lmdb_bytes().to_vec())
+        {
+            Some(val_buf) => match bincode::deserialize(val_buf) {
+                Ok(val) => Ok(Some(val)),
+                Err(_) => Ok(None),
+            },
+            None => Ok(None),
+        }
+    }
+
+    /// Returns the total number of entries recorded in the key-value database.
+    pub fn len(&self) -> Result<usize, KeyValStoreError> {
+        Ok(self.db.read().map_err(|_| KeyValStoreError::PoisonedAccess)?.len())
+    }
+
+    /// Iterate over all the stored records and execute the function `f` for each pair in the key-value database.
+    pub fn for_each<K, V, F>(&self, mut f: F) -> Result<(), KeyValStoreError>
+    where
+        K: serde::de::DeserializeOwned,
+        V: serde::de::DeserializeOwned,
+        F: FnMut(Result<(K, V), KeyValStoreError>),
+    {
+        for (key_buf, val_buf) in self
+            .db
+            .read()
+            .map_err(|_| KeyValStoreError::PoisonedAccess)?
+            .clone()
+            .into_iter()
+        {
+            let key =
+                bincode::deserialize(&key_buf).map_err(|e| KeyValStoreError::DeserializationError(e.to_string()))?;
+            let val =
+                bincode::deserialize(&val_buf).map_err(|e| KeyValStoreError::DeserializationError(e.to_string()))?;
+
+            f(Ok((key, val)));
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether a record exist in the key-value database that corresponds to the provided `key`.
+    pub fn contains_key<K>(&self, key: &K) -> Result<bool, KeyValStoreError>
+    where K: AsLmdbBytes + ?Sized {
+        Ok(self
+            .db
+            .read()
+            .map_err(|_| KeyValStoreError::PoisonedAccess)?
+            .contains_key(&key.as_lmdb_bytes().to_vec()))
+    }
+
+    /// Remove the record from the key-value database that corresponds with the provided `key`.
+    pub fn remove<K>(&self, key: &K) -> Result<(), KeyValStoreError>
+    where K: AsLmdbBytes + ?Sized {
+        let _ = self
+            .db
+            .write()
+            .map_err(|_| KeyValStoreError::PoisonedAccess)?
+            .remove(&key.as_lmdb_bytes().to_vec());
```

Comment:
This will return Ok on the key not existing, is this intended behavior?

Reply:
It was intended, but on closer inspection you are right. LMDB returns a `NOTFOUND` if the key does not exist. The HashMap only returns None so I will change the behaviour to return an Error.